test/recipes/82-test_ocsp_cert_chain.t

```diff
@@ -0,0 +1,147 @@
+    my $r = $A + int(rand($B - $A));
+    return $r;
+}
+
+sub run_test {
+
+    # this test starts two servers that listen on respective ports.
+    # that can be problematic since the ports may not be available
+    # (e.g. when multiple instances of the test are run on the same
+    # machine).
+
+    # we use random ports to help avoid this.
+
+    # using a random port with s_server is straightforward.  doing so
+    # with the ocsp responder required some investigation because the
+    # url for the ocsp responder is usually included in the server's
+    # cert (normally, in the authority information access extension,
+    # and it would be complicated to change that when the test
+    # executes).  however, s_server has an option "-status_url" that
+    # can be used to specify a fallback url when no url is specified
+    # in the cert.  that is what we do here.
+
+    my $ocsp_port = randrange(19200, 19300);
+    my $server_port = randrange(4400, 4500);
+
+    # openssl ocsp -port 19254 -index index.txt -rsigner ocsp.pem -CA intermediate-cert.pem
+    my @ocsp_cmd = ("ocsp", "-port", "$ocsp_port", "-index", $index_txt, "-rsigner", $ocsp_pem, "-CA", $intermediate_cert_pem);
+    my $ocsp_pid = open3(my $ocsp_i, my $ocsp_o, my $ocsp_e = gensym, $shlib_wrap, $apps_openssl, @ocsp_cmd);
+
+    ## ipv4
+    # ACCEPT 0.0.0.0:19254 PID=620007
+    ## ipv6
+    # ACCEPT [::]:19254 PID=620007
+    my $port = "0";
+    while (<$ocsp_o>) {
+        print($_);
+        chomp;
+        if (/^ACCEPT 0.0.0.0:(\d+)/) {
+            $port = $1;
+            last;
+        } elsif (/^ACCEPT \[::\]:(\d+)/) {
+            $port = $1;
+            last;
+        } else {
+            last;
+        }
+    }
+    ok($port eq $ocsp_port, "ocsp server port check");
+
+    print("ocsp server ready, listening on port $port\n");
+
+    # openssl s_server -accept 4433 -cert server.pem -cert_chain intermediate-cert.pem \
+    #                  -status_verbose -status_url http://127.0.0.1:19254/ocsp
+    my @s_server_cmd = ("s_server", "-accept", "$server_port", "-cert", $server_pem, "-cert_chain", $intermediate_cert_pem,
+                        "-status_verbose", "-status_url", "http://127.0.0.1:${ocsp_port}/ocsp");
+    my $s_server_pid = open3(my $s_server_i, my $s_server_o, my $s_server_e = gensym, $shlib_wrap, $apps_openssl, @s_server_cmd);
+
+    # ACCEPT
+    my $server_ready = 0;
+    while (<$s_server_o>) {
+        print($_);
+        chomp;
+        if (/^Using default/) {
+            ;
+        } elsif (/^ACCEPT$/) {
+            $server_ready = 1;
+            last;
+        } else {
+            last;
+        }
+    }
+    ok($server_ready == 1, "s_server ready check");
+
+    print("s_server ready, listening on port $server_port\n");
+
+    # openssl s_client -connect 0.0.0.0:4433 -status -verify_return_error
+    my @s_client_cmd = ("s_client", "-connect", "0.0.0.0:$server_port", "-status", "-verify_return_error");
```

Comment:
??? I'm not sure what 0.0.0.0 means from a client perspective...in any case it will cause problems on an IPv6 only machine.

Reply:
on linux, `0.0.0.0` can apparently be used like `localhost`:

https://en.wikipedia.org/wiki/0.0.0.0

It is probably better to use `localhost` instead.